### PR TITLE
Micro optimizations of the virtual machine

### PR DESCRIPTION
**Pull request by zherczeg**
- Branch argument information is encoded in the vm byte
    code data, so CBC flags are not loaded anymore
  - The free_flags variable is removed from the vm_loop
  - Fast code paths are added for free_value and copy_value
  - Two cases are removed from "get arguments", argument
    processing is simplified
  - A dead code is removed: collections are not used anymore
    for passing arguments. This code path accidentally remanined
    when the support was removed.

### CODE REVIEW

**Comment by zherczeg** [comment]
Improves all test cases:

Benchmark | Perf(sec) |
---- | ---- | 
3d-cube.js | 2.101 -> 2.012 : 4.22% | 
3d-raytrace.js | 2.254 -> 2.191 : 2.82% | 
access-binary-trees.js | 1.136 -> 1.098 : 3.35% | 
access-fannkuch.js | 5.663 -> 5.285 : 6.69% | 
access-nbody.js | 2.443 -> 2.354 : 3.62% | 
bitops-3bit-bits-in-byte.js | 0.986 -> 0.845 : 14.29% | 
bitops-bits-in-byte.js | 1.344 -> 1.151 : 14.35% | 
bitops-bitwise-and.js | 2.650 -> 2.360 : 10.95% | 
bitops-nsieve-bits.js | 3.572 -> 3.252 : 8.93% | 
controlflow-recursive.js | 0.664 -> 0.588 : 11.33% | 
crypto-aes.js | 2.638 -> 2.516 : 4.62% | 
crypto-md5.js | 1.133 -> 1.063 : 6.13% | 
crypto-sha1.js | 1.065 -> 0.987 : 7.32% | 
date-format-tofte.js | 1.655 -> 1.552 : 6.22% | 
date-format-xparb.js | 0.707 -> 0.667 : 5.68% | 
math-cordic.js | 2.427 -> 2.179 : 10.20% | 
math-partial-sums.js | 1.374 -> 1.329 : 3.28% | 
math-spectral-norm.js | 1.125 -> 1.027 : 8.69% | 
string-base64.js | 5.074 -> 4.964 : 2.16% | 
string-fasta.js | 2.695 -> 2.571 : 4.61% | 

Binary sizes (bytes)
e1c8a2ec96:168524
77a07e5325:168784

**Comment by zherczeg** [comment]
Updated results

Benchmark | Perf(sec) |
---- | ---- | 
3d-cube.js | 2.101 -> 2.007 : 4.46% | 
3d-raytrace.js | 2.254 -> 2.182 : 3.21% | 
access-binary-trees.js | 1.136 -> 1.093 : 3.78% | 
access-fannkuch.js | 5.663 -> 4.946 : 12.67% | 
access-nbody.js | 2.443 -> 2.305 : 5.61% | 
bitops-3bit-bits-in-byte.js | 0.986 -> 0.852 : 13.59% | 
bitops-bits-in-byte.js | 1.344 -> 1.174 : 12.67% | 
bitops-bitwise-and.js | 2.650 -> 2.320 : 12.46% | 
bitops-nsieve-bits.js | 3.572 -> 3.232 : 9.52% | 
controlflow-recursive.js | 0.664 -> 0.594 : 10.47% | 
crypto-aes.js | 2.638 -> 2.445 : 7.33% | 
crypto-md5.js | 1.133 -> 1.061 : 6.36% | 
crypto-sha1.js | 1.065 -> 0.990 : 7.09% | 
date-format-tofte.js | 1.655 -> 1.562 : 5.67% | 
date-format-xparb.js | 0.707 -> 0.669 : 5.48% | 
math-cordic.js | 2.427 -> 2.195 : 9.55% | 
math-partial-sums.js | 1.374 -> 1.278 : 7.02% | 
math-spectral-norm.js | 1.125 -> 1.058 : 5.90% | 
string-base64.js | 5.074 -> 4.941 : 2.60% | 
string-fasta.js | 2.695 -> 2.588 : 3.96% | 

Binary sizes (bytes)
e1c8a2ec96:168524
20fcc7dabb:168212

File: jerry-core/vm/vm.c [hunk 1]
```diff
@@ -794,92 +769,63 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
...
         {
           uint16_t literal_index;
           READ_LITERAL_INDEX (literal_index);
           READ_LITERAL (literal_index,
-                        left_value,
-                        free_flags = VM_FREE_LEFT_VALUE);
+                        left_value);
 
-          switch (operands)
+          if (operands != VM_OC_GET_LITERAL)
```
Comment (akiss77) [inline jerry-core/vm/vm.c]: Why did you change for an `if () { if () {} else if () {} else {} }` construct?
```c
switch (operands)
{
  case VM_OC_GET_LITERAL:
    break;
  case VM_OC_GET_LITERAL_LITERAL:
    ...
  case VM_VM_OC_GET_STACK_LITERAL:
    ...
  default:
    ...
}
```
The above does exactly the same, is more readable (although that's a subjective matter), and I don't see how this would be less performant.

File: jerry-core/ecma/base/ecma-helpers-value.c [hunk 2]
```diff
@@ -615,6 +615,27 @@ ecma_copy_value (ecma_value_t value)  /**< value description */
...
+ * @return copy of the given value
+ */
+inline ecma_value_t __attr_always_inline___
+ecma_fast_copy_value (ecma_value_t value)  /**< value description */
+{
+  if (ecma_get_value_type_field (value) != ECMA_TYPE_DIRECT)
+  {
+    value = ecma_copy_value (value);
+  }
+  return value;
```
Comment (akiss77) [inline jerry-core/ecma/base/ecma-helpers-value.c]: What about
```c
return ecma_get_value_type_field (value) == ECMA_TYPE_DIRECT ? value : ecma_copy_value (value);
```

Or, if you prefer an `if` construct, let's use an early return approach at least:
```c
if (ecma_get_value_type_field (value) != ECMA_TYPE_DIRECT)
{
  return ecma_copy_value (value);
}
return value;
```

File: jerry-core/vm/vm.c (hunk 1 above)
Comment (zherczeg) [inline jerry-core/vm/vm.c]: Why dioes it more readable? I also need to add a lot of extra {} around cases, it is too big.

File: jerry-core/ecma/base/ecma-helpers-value.c (hunk 2 above)
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-value.c]: I was hoping this helps for the compiler. Today I analysed a code where the compiler pushed a value to the memory and read it back because it does not understand such a simple construct.

File: jerry-core/vm/vm.c (hunk 1 above)
Comment (akiss77) [inline jerry-core/vm/vm.c]: Because all the predicates are equality checks on the same variable, `operands`. That's the typical use case for a switch.
The braces around the cases are no extras, they are already there for the if/if-else/else branches. (To be fair: some extra comes from the breaks and the case for `VM_OC_GET_LITERAL`.)

Anyway, I just wanted to point out that nesting `if`s with equality checks on the same var is not that nice. But also mentioned that it's a matter of taste.

File: jerry-core/ecma/base/ecma-helpers-value.c (hunk 2 above)
Comment (akiss77) [inline jerry-core/ecma/base/ecma-helpers-value.c]: I'm not sure I get this. You mean that you had an example (in this very function?) where a compiler generated more efficient code for `if (p) { a = f(a); } return a;` than for `return p ? f(a) : a;`? That's against all my experience.

**Comment by zherczeg** [comment]
Several parts got an own patch, this patch is shrinked now.

**Comment by LaszloLango** [comment]
LGTM

File: jerry-core/vm/vm.c [hunk 3]
```diff
@@ -794,92 +769,70 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
...
-            case VM_OC_GET_ARGS_GET_INDEX (VM_OC_GET_STACK_LITERAL):
-            {
-              JERRY_ASSERT (stack_top_p > frame_ctx_p->registers_p + register_end);
-              right_value = left_value;
-              left_value = *(--stack_top_p);
-              free_flags = (uint8_t) ((free_flags << 1) | VM_FREE_LEFT_VALUE);
-              break;
-            }
-            case VM_OC_GET_ARGS_GET_INDEX (VM_OC_GET_LITERAL_BYTE):
+            case VM_OC_GET_LITERAL:
```
Comment (akiss77) [inline jerry-core/vm/vm.c]: Is there a reason why the case branches have been shuffled? I'd leave the no-op `VM_OC_GET_LITERAL` case as the default.

Comment (zherczeg) [inline jerry-core/vm/vm.c]: Yes. The VM_OC_GET_LITERAL has the highest chance, so it would be great if that would be checked first. The original "if" statement did it as well.

Comment (akiss77) [inline jerry-core/vm/vm.c]: I'm not sure you can control switch layout at this level. I just built this PR and on x86_64 in release mode, and got the following:
```asm
  402b17:       66 81 7c 24 50 00 02    cmpw   $0x200,0x50(%rsp) ; VM_OC_GET_LITERAL_LITERAL
  402b1e:       74 35                   je     402b55 <vm_loop+0x2c1>
  402b20:       66 81 7c 24 50 80 02    cmpw   $0x280,0x50(%rsp) ; VM_OC_GET_STACK_LITERAL
  402b27:       0f 84 35 01 00 00       je     402c62 <vm_loop+0x3ce>
  402b2d:       66 81 7c 24 50 80 01    cmpw   $0x180,0x50(%rsp) ; VM_OC_GET_LITERAL
  402b34:       c7 04 24 18 00 00 00    movl   $0x18,(%rsp)
  402b3b:       0f 84 8a 01 00 00       je     402ccb <vm_loop+0x437>
```
The compiler generated code to check for `VM_OC_GET_LITERAL` as the third step. Is `VM_OC_GET_LITERAL` really that frequent to make it worth factor out into an enclosing `if`?

**Comment by zherczeg** [comment]
Patch updated

**Comment by akiss77** [comment]
LGTM, after the CI is flipped back to green

**Comment by zherczeg** [comment]
Checks are passing. I think the previously landed bugfix fixed those flaky tests.
